File: nvvk/extensions_vk.lua
--[[
/*
 * Copyright (c) 2018-2021, NVIDIA CORPORATION.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-FileCopyrightText: Copyright (c) 2018-2021 NVIDIA CORPORATION
 * SPDX-License-Identifier: Apache-2.0
 */

]]

-- HOW TO USE
--
-- 1. Setup environment variable NVVK_VULKAN_XML pointing to vk.xml
--    or use VULKAN_SDK >= 1.2.135.0
--
-- 2. Modify the enablelist of extensions you may want to use.
--    Only those extensions that have functions will get a loader
--    (load_VK_BLAH_whatever) so not all extensions in this list
--    may get a function in the end.
--
-- 3. Check out this and the other extensions_vk files for write access
--
-- 4. Run with a lua5.1 compatible lua runtime and the lua2xml project
--    https://github.com/manoelcampos/xml2lua
--    (shared_internal has all the files).
--
--    lua extensions_vk.lua
--
--    within this directory.

local VULKAN_XML = os.getenv("NVVK_VULKAN_XML") or os.getenv("VULKAN_SDK").."/share/vulkan/registry/vk.xml"
local extensionSubset = [[
    VK_KHR_ray_tracing
    VK_KHR_push_descriptor
    VK_KHR_8bit_storage
    VK_KHR_create_renderpass2
    VK_KHR_depth_stencil_resolve
    VK_KHR_draw_indirect_count
    VK_KHR_driver_properties
    VK_KHR_pipeline_executable_properties
    VK_KHR_buffer_device_address
    
    VK_NV_compute_shader_derivatives
    VK_NV_cooperative_matrix
    VK_NV_corner_sampled_image
    VK_NV_coverage_reduction_mode
    VK_NV_dedicated_allocation_image_aliasing
    VK_NV_mesh_shader
    VK_NV_ray_tracing
    VK_NV_representative_fragment_test
    VK_NV_shading_rate_image
    VK_NV_viewport_array2
    VK_NV_viewport_swizzle
    VK_NV_scissor_exclusive
    VK_NV_device_generated_commands
    
    VK_EXT_buffer_device_address
    VK_EXT_debug_marker
    VK_EXT_calibrated_timestamps
    VK_EXT_conservative_rasterization
    VK_EXT_descriptor_indexing
    VK_EXT_depth_clip_enable
    VK_EXT_memory_budget
    VK_EXT_memory_priority
    VK_EXT_pci_bus_info
    VK_EXT_sample_locations
    VK_EXT_sampler_filter_minmax
    VK_EXT_texel_buffer_alignment
    VK_EXT_debug_utils
    VK_EXT_host_query_reset
    
    VK_KHR_external_memory_win32
    VK_KHR_external_semaphore_win32
    VK_KHR_external_fence_win32

    VK_KHR_external_memory_fd
    VK_KHR_external_semaphore_fd
	
	VK_KHR_acceleration_structure
    VK_KHR_ray_tracing_pipeline
	VK_KHR_pipeline_library
    VK_KHR_synchronization2
    ]]

local function extractFeatureDefs(features, filename, enablelist)
  assert(filename, "filename not provided")
  
  local xml2lua = require("xml2lua")
  local handler = require("xmlhandler.tree")
  
  local f = io.open(filename,"rt")
  assert(f, filename.." not found")
  
  local xml = f:read("*a")
  f:close()
  
  -- Bug workaround https://github.com/manoelcampos/xml2lua/issues/35
  xml = xml:gsub("(<param>)(<type>[%w_]+</type>)%* ", function(p,typ)
        -- add _ dummy symbol
        return "<param>_"..typ.."* "
      end)
    
  local parser = xml2lua.parser(handler)
  parser:parse(xml)
  
  local version = xml:match("VK_HEADER_VERSION</name> (%d+)")
  assert(version)
  
  xml = nil

  --[[
          <command>
              <proto><type>VkDeviceAddress</type> <name>vkGetBufferDeviceAddress</name></proto>
              <param><type>VkDevice</type> <name>device</name></param>
              <param>const <type>VkBufferDeviceAddressInfo</type>* <name>pInfo</name></param>
          </command>
          <command name="vkGetBufferDeviceAddressKHR"        alias="vkGetBufferDeviceAddress"/>
          <command name="vkGetBufferDeviceAddressEXT"        alias="vkGetBufferDeviceAddress"/>
          
          
          <extension name="VK_EXT_buffer_device_address" number="245" type="device" requires="VK_KHR_get_physical_device_properties2" author="NV" contact="Jeff Bolz @jeffbolznv"  deprecatedby="VK_KHR_buffer_device_address" supported="vulkan">
              <require>
                  ...
                  <command name="vkGetBufferDeviceAddressEXT"/>
              </require>
              
          <extension name="VK_KHR_win32_surface" number="10" type="instance" requires="VK_KHR_surface" platform="win32" author="KHR" contact="Jesse Hall @critsec,Ian Elliott @ianelliottus" supported="vulkan">
              <require>
                  ...
                  <command name="vkCreateWin32SurfaceKHR"/>
                  <command name="vkGetPhysicalDeviceWin32PresentationSupportKHR"/>
              </require>
          </extension>
  ]]
  
  local types      = handler.root.registry.types
  local commands   = handler.root.registry.commands
  local extensions = handler.root.registry.extensions.extension
  
  -- debugging
  if (false) then
    local serpent = require "serpent"
    local f = io.open(filename..".cmds.lua", "wt")
    f:write(serpent.block(commands))
    local f = io.open(filename..".exts.lua", "wt")
    f:write(serpent.block(extensions))
  end
  
  -- build list of alias'ed types
  local lktypes = {}
  for _,v in ipairs(types.type) do
    local alias = v._attr.alias
    if (alias) then
      lktypes[v._attr.name] = alias
    end
  end

  -- build list of commands, and aliased commands
  local lkcmds = {}
  for _,v in ipairs(commands.command) do
    if (v.proto) then
      lkcmds[v.proto.name] = v
    elseif (v._attr.alias) then
      lkcmds[v._attr.name] = lkcmds[v._attr.alias]
      assert(lkcmds[v._attr.name].proto.name)
    end
  end
  
  
  local platforms = {
    ggp = "VK_USE_PLATFORM_GGP",
    win32 = "VK_USE_PLATFORM_WIN32_KHR",
    vi = "VK_USE_PLATFORM_VI_NN",
    ios = "VK_USE_PLATFORM_IOS_MVK",
    macos = "VK_USE_PLATFORM_MACOS_MVK",
    android = "VK_USE_PLATFORM_ANDROID_KHR",
    fuchsia = "VK_USE_PLATFORM_FUCHSIA",
    metal = "VK_USE_PLATFORM_METAL_EXT",
    xlib = "VK_USE_PLATFORM_XLIB_KHR",
    xcb = "VK_USE_PLATFORM_XCB_KHR",
    wayland = "VK_USE_PLATFORM_WAYLAND_KHR",
    xlib_xrandr = "VK_USE_PLATFORM_XLIB_XRANDR_EXT",
  }
  for _,v in ipairs(extensions) do
    local extname = v._attr.name
    local extrequires = v.require
    if (extrequires and not extrequires[1]) then
      extrequires = {extrequires}
    end
    
    if (enablelist[extname]) then
      local alias = {}
      local cmds = {}
      local hascmds = false
      local addedcmds = {}
      for _,r in ipairs(extrequires) do
        local extcmds = r.command
        local exttypes = r.type
        if (extcmds) then
          -- convert single entry to array
          if (not extcmds[1]) then
            extcmds = {extcmds}
          end
          -- extract commands used by extension
          for _,c in ipairs(extcmds) do
            local cname = c._attr.name
            local cmd = lkcmds[cname]
            
            assert(cmd, extname..":"..cname)
            if (not addedcmds[cname]) then
              table.insert(cmds, {name=cname, cmd=cmd})
              addedcmds[cname] = true
            end
          end
          -- extract type aliasing 
          -- we prefer to use the extensions original types
          -- which may have been replaced through aliasing
          if (exttypes) then
            if (not exttypes[1]) then
              exttypes = {exttypes}
            end
            for _,t in ipairs(exttypes) do
              local tname = t._attr.name
              local aname = lktypes[tname]
              if (aname) then
                alias[aname] = tname
              end
            end
          end
        end
      end
      if (cmds[1]) then
        print(extname)
        table.insert(features, {feature=extname, typ=v._attr.type, cmds=cmds, alias=alias, platform=platforms[v._attr.platform or "_"]})
      end
    end
  end
  
  return version
end

local function generate(outfilename, header, enablelist)

  local function toTab(str)
    local tab = {}
    for name in str:gmatch("[%w_]+") do 
      tab[name] = true
    end
    return tab
  end
  
  local enablelist = enablelist and toTab(enablelist)

  local xmlfile = VULKAN_XML
  
  local vk_features = {}
  local vk_version = extractFeatureDefs(vk_features, xmlfile, enablelist)
  
  local availables_header = ""
  local availables_source = ""
  
  local loaders_header = ""
  local loaders_source = ""
  local loaders_all = ""
  local reset_all = ""
  
  local function process(f, api, apientry) 
    local func      = ""
    local initfunc  = ""
    local success   = ""
    local variable  = ""
    local reset     = ""
    
    for _,c in ipairs(f.cmds) do
      assert(c.cmd and c.name, f.feature)
      local cmd = c.cmd
      local name = c.name
      local dbgname = f.feature..":"..name
      
      local prot = "PFN_"..name
      local mangled = "pfn_"..name
      local ret = cmd.proto.type
      
      local params = cmd.param
      assert(params, dbgname)
      if (not (params[1] and params[1].type))  then
        params = {params}
      end
      
      local isInstance = params[1].type == "VkInstance"
      
      -- argument definition
      local args = "(\n"
      -- argument passing
      local exec = (ret ~= "void" and "return " or "")..mangled.."("
      
      for _,p in ipairs(params) do
        assert(p.type and p.name, dbgname)
        -- convert promoted types back to extension types
        local typ = f.alias[p.type] or p.type
        -- handle qualifiers
        local p1 = p[1] == "const" and "const " or ""
        local function pointer(a,b)
          --local convert = {["*"] = "*", ["**"] = "**", ["* const*"] = "* const*",}
          return (a or b or ""):find("*",0, true) and (a or b) or ""
        end
        local p2 = pointer(p[2],p[1])
        args = args.."    "..p1..typ..p2.." "..p.name..",\n"
        exec = exec..p.name..","
      end
      
      args = args:sub(1,-3)..")"
      exec = exec:sub(1,-2)..");"
      
      variable = variable.."static "..prot.." "..mangled.." = 0;\n"
      reset    = reset.."  "..prot.." "..mangled.." = 0;\n"
      
      func = func..api..ret.." "..apientry..name..args.."\n{\n"
      func = func.."  assert("..mangled..");\n"
      func = func.."  "..exec.."\n"
      func = func.."}\n"
      
      if (isInstance) then
        initfunc  = initfunc.."  "..mangled..' = ('..prot..')getInstanceProcAddr(instance, "'..name..'");\n'
      else
        initfunc  = initfunc.."  "..mangled..' = ('..prot..')getDeviceProcAddr(device, "'..name..'");\n'
      end
      success  = success.."  success = success && ("..mangled..' != 0);\n'
    end
    
    if (initfunc ~= "") then
      local preproc_condition = "#if "..f.feature..(f.platform and " && defined("..f.platform..")" or "").."\n"
      
      loaders_header = loaders_header..preproc_condition
      loaders_header = loaders_header.."#define LOADER_"..f.feature.." 1\n"
      loaders_header = loaders_header.."int load_"..f.feature.."(VkInstance instance, PFN_vkGetInstanceProcAddr getInstanceProcAddr, VkDevice device, PFN_vkGetDeviceProcAddr getDeviceProcAddr);\n"
      loaders_header = loaders_header.."extern int has_"..f.feature..";\n"
      loaders_header = loaders_header.."#endif\n\n"
      
      loaders_source = loaders_source.."/*
 * Copyright (c) 2018-2021, NVIDIA CORPORATION.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-FileCopyrightText: Copyright (c) 2018-2021 NVIDIA CORPORATION
 * SPDX-License-Identifier: Apache-2.0
 */


//////////////////////////////////////////////////////////////////////////
/*
 * Copyright (c) 2018-2021, NVIDIA CORPORATION.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-FileCopyrightText: Copyright (c) 2018-2021 NVIDIA CORPORATION
 * SPDX-License-Identifier: Apache-2.0
 */


#include <assert.h>
#include "]]..outfilename..[[.hpp"

/* loaders */
]]..loaders_source..[[

/* super load/reset */
void load_VK_EXTENSION_SUBSET(VkInstance instance, PFN_vkGetInstanceProcAddr getInstanceProcAddr, VkDevice device, PFN_vkGetDeviceProcAddr getDeviceProcAddr) {
]]..loaders_all..[[
}
void reset_VK_EXTENSION_SUBSET() {
]]..reset_all..[[
}

]])




end

do
  generate("extensions_vk",
    "/* auto generated by extensions_vk.lua */",
    extensionSubset)
end

File: nvvk/extensions_vk.py
#
#  Copyright (c) 2018-2021, NVIDIA CORPORATION.  All rights reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
#  SPDX-FileCopyrightText: Copyright (c) 2018-2021 NVIDIA CORPORATION
#  SPDX-License-Identifier: Apache-2.0
#

#
# The following script originate from Volk (https://github.com/zeux/volk) and adapted to the need
# of nvpro-core samples.
#


# Copyright (c) 2018-2023 Arseny Kapoulkine
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:


#!/usr/bin/python3
# This will generate the entry point for all Vulkan extensions
# Code blocks are created and will be replace between
# 'NVVK_GENERATE_'<BLOCK_NAME>
import argparse
import os.path
import urllib
import urllib.request
import xml.etree.ElementTree as etree
import re
from collections import OrderedDict

# Ignoring those extensions because they are part of vulkan-1.lib
ExcludeList = [
    "defined(VK_KHR_surface)",
    "defined(VK_KHR_win32_surface)",
    "defined(VK_KHR_xlib_surface)",
    "defined(VK_KHR_wayland_surface)",
    "defined(VK_KHR_xcb_surface)",
    "defined(VK_KHR_display)",
    "defined(VK_KHR_swapchain)",
    "defined(VK_KHR_get_surface_capabilities2)",
    "defined(VK_KHR_get_display_properties2)",
    "defined(VK_KHR_display_swapchain)",
    "VK_VERSION_1_0",
    "VK_VERSION_1_1",
    "VK_VERSION_1_2",
    "VK_VERSION_1_3",
]

# Debugging - To be sure that the exclude list is excluding commands
# exported in vulkan-1, populate the list here. If there is a duplicate
# the name of the command and the extension name will be printed out.
ExportedCommands = []  # dumpbin /EXPORTS vulkan-1.lib


# Commands that were added in newer extension revisions.
# Extensions such as VK_EXT_discard_rectangles have had specification revisions
# that added new commands. Since these commands should only be used if the
# extension's `VkExtensionProperties::specVersion` is high enough, this table
# tracks the first `specVersion` in which each newer command was introduced
# (as this information is not currently contained in vk.xml).
cmdversions = {
    "vkCmdSetDiscardRectangleEnableEXT": 2,
    "vkCmdSetDiscardRectangleModeEXT": 2,
    "vkCmdSetExclusiveScissorEnableNV": 2,
}


def parse_xml(path):
    # Parsing the Vulkan 'vk.xml' document
    file = urllib.request.urlopen(path) if path.startswith("http") else open(path, "r")
    with file:
        tree = etree.parse(file)
        return tree


def patch_file(fileName, blocks):
    # Find each section of NVVK_GENERATE_ and replace with block of text
    result = []
    block = None

    scriptDir = os.path.dirname(os.path.realpath(__file__))
    path = os.path.join(scriptDir, fileName)

    with open(path, "r") as file:
        for line in file.readlines():
            if block:
                if line == block:
                    result.append(line)
                    block = None
            else:
                result.append(line)
                # C comment marker
                if line.strip().startswith("/* NVVK_GENERATE_"):
                    block = line
                    result.append(blocks[line.strip()[17:-3]])

    with open(path, "w", newline="\n") as file:
        for line in result:
            file.write(line)


def is_descendant_type(types, name, base):
    # Finding the base type of each type:
    # <type category="handle" parent="VkPhysicalDevice"
    # objtypeenum="VK_OBJECT_TYPE_DEVICE"><type>VK_DEFINE_HANDLE</type>(<name>VkDevice</name>)</type>
    # <type category="handle" parent="VkDevice"
    # objtypeenum="VK_OBJECT_TYPE_QUEUE"><type>VK_DEFINE_HANDLE</type>(<name>VkQueue</name>)</type>
    if name == base:
        return True
    type = types.get(name)
    if not type:
        return False
    parents = type.get("parent")
    if not parents:
        return False
    return any(
        [is_descendant_type(types, parent, base) for parent in parents.split(",")]
    )


def defined(key):
    return "defined(" + key + ")"


def cdepends(key):
    return (
        re.sub(r"[a-zA-Z0-9_]+", lambda m: defined(m.group(0)), key)
        .replace(",", " || ")
        .replace("+", " && ")
    )


# Remove "defined(..)"
def remove_defined(input_string):
    return re.sub(r"defined\((.*?)\)", r"\1", input_string)

def toStr(txt):
    # Return the string if it exist or '' if None
    if txt:
        return txt
    return ""


def get_function(rtype, name, params):
    # Returning the function declaration

    fct_args = []  # incoming argument
    call_args = []  # call arguments
    for p in params:
        ptype = p.find("type")
        pname = p.find("name")
        papi = p.attrib.get("api")
        # Avoid `vulkansc`
        if not papi or papi == "vulkan":
            fct_args.append(
                "".join(
                    [
                        toStr(p.text),
                        ptype.text,
                        ptype.tail,
                        pname.text,
                        toStr(pname.tail),
                    ]
                )
            )  # 'const', 'vkSome', '*', 'some', '[2]'
            call_args.append(pname.text)

    # Function signature
    fct = "VKAPI_ATTR " + rtype + " VKAPI_CALL " + name + "(\n"
    # Arguments of the function
    fct += "\t" + ", \n\t".join(fct_args) + ") \n"
    fct += "{ \n  "
    # fct += ' assert(pfn_'+name+');\n'
    # Check if the function is returning a value
    if rtype != "void":
        fct += "return "
    fct += "pfn_" + name + "(" + ", ".join(call_args) + "); \n"
    fct += "}\n"
    return fct


def get_vk_xml_path(spec_arg):
    """
    Find the Vulkan specification XML file by looking for (highest priority to
    lowest) an incoming `spec` argument, a local copy within the Vulkan SDK,
    or by downloading it from KhronosGroup/Vulkan-Docs.
    """
    if spec_arg is not None:
        return spec_arg

    # VULKAN_SDK is a newer version of VK_SDK_PATH. The Linux Tarball Vulkan SDK
    # instructions only say to set VULKAN_SDK - so VULKAN_SDK might exist while
    # VK_SDK_PATH might not.
    vulkan_sdk_env = os.getenv("VULKAN_SDK")
    if vulkan_sdk_env is not None:
        local_spec_path = os.path.normpath(
            vulkan_sdk_env + "/share/vulkan/registry/vk.xml"
        )
        if os.path.isfile(local_spec_path):
            return local_spec_path

    # Ubuntu installations might not have VULKAN_SDK set, but have vk.xml in /usr.
    if os.path.isfile("/usr/share/vulkan/registry/vk.xml"):
        return "/usr/share/vulkan/registry/vk.xml"

    print(
        "Warning: no `spec` parameter was provided, and vk.xml could not be "
        "found in the path given by the VULKAN_SDK environment variable or in "
        "system folders. This script will download the latest copy of vk.xml "
        "online, which may be incompatible with an installed Vulkan installation."
    )
    return "https://raw.githubusercontent.com/KhronosGroup/Vulkan-Docs/main/xml/vk.xml"


#
# MAIN Entry
#
if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Generates entry points for Vulkan extensions in extensions_vk.cpp."
    )
    parser.add_argument(
        "--beta",
        action="store_true",
        help="Includes provisional Vulkan extensions; these extensions are not guaranteed to be consistent across Vulkan SDK versions.",
    )
    parser.add_argument(
        "spec",
        type=str,
        nargs="?",
        help="Optional path to a vk.xml file to use to generate extensions. Otherwise, uses the vk.xml in the Vulkan SDK distribution specified in the VULKAN_SDK environment variable.",
    )

    args = parser.parse_args()

    # Retrieving the XML file
    specpath = get_vk_xml_path(args.spec)
    spec = parse_xml(specpath)

    # CODE BLOCS
    blocks = {}

    # CODE BLOCS for generated code
    block_keys = ("STATIC_PFN", "LOAD_PROC", "DECLARE", "DEFINE")

    for key in block_keys:
        blocks[key] = ""

    # Retrieving the version of the Vulkan specification
    version = spec.find('types/type[name="VK_HEADER_VERSION"]')
    blocks["VERSION_INFO"] = (
        "// Generated using Vulkan " + version.find("name").tail.strip() + "\n"
    )

    # Patching the files
    patch_file("extensions_vk.cpp", blocks)

    # Ordered list of commands per extension group
    command_groups = OrderedDict()
    instance_commands = set()

    for feature in spec.findall("feature"):
        api = feature.get("api")
        if "vulkan" not in api.split(","):
            continue
        key = feature.get("name")
        cmdrefs = feature.findall("require/command")
        command_groups[key] = [cmdref.get("name") for cmdref in cmdrefs]

    # Retrieve all extension and sorted alphabetically
    for ext in sorted(
        spec.findall("extensions/extension"), key=lambda ext: ext.get("name")
    ):
        # Only add the extension if 'vulkan' is part of the support attribute
        supported = ext.get("supported")
        if "vulkan" not in supported.split(","):
            continue

        # Discard beta extensions
        if ext.get("provisional") == "true" and not args.beta:
            continue

        name = ext.get("name")
        type = ext.get("type")  # device or instance

        for req in ext.findall("require"):
            # Adding all commands for this extension
            key = defined(name)
            if req.get("feature"):  # old-style XML depends specification
                for i in req.get("feature").split(","):
                    key += " && " + defined(i)
            if req.get("extension"):  # old-style XML depends specification
                for i in req.get("extension").split(","):
                    key += " && " + defined(i)
            if req.get("depends"):  # new-style XML depends specification
                dep = cdepends(req.get("depends"))
                key += " && " + ("(" + dep + ")" if "||" in dep else dep)
            cmdrefs = req.findall("command")

            # Add ifdef section and split commands with high version
            for cmdref in cmdrefs:
                ver = cmdversions.get(cmdref.get("name"))
                if ver:
                    command_groups.setdefault(
                        key + " && " + name.upper() + "_SPEC_VERSION >= " + str(ver), []
                    ).append(cmdref.get("name"))
                else:
                    command_groups.setdefault(key, []).append(cmdref.get("name"))

            # Adding commands that are 'instance' instead of 'device'
            if type == "instance":
                for cmdref in cmdrefs:
                    instance_commands.add(cmdref.get("name"))

    # From a command, find which group it's belong
    commands_to_groups = OrderedDict()
    for group, cmdnames in command_groups.items():
        for name in cmdnames:
            commands_to_groups.setdefault(name, []).append(group)

    for group, cmdnames in command_groups.items():
        command_groups[group] = [
            name for name in cmdnames if len(commands_to_groups[name]) == 1
        ]

    for name, groups in commands_to_groups.items():
        if len(groups) == 1:
            continue
        key = " || ".join([g for g in groups])
        command_groups.setdefault(key, []).append(name)

    # Finding the alias name for a function: <command
    # name="vkGetPhysicalDeviceExternalBufferPropertiesKHR"
    # alias="vkGetPhysicalDeviceExternalBufferProperties"/>
    commands = {}
    for cmd in spec.findall("commands/command"):
        if not cmd.get("alias"):
            name = cmd.findtext("proto/name")
            commands[name] = cmd

    for cmd in spec.findall("commands/command"):
        if cmd.get("alias"):
            name = cmd.get("name")
            commands[name] = commands[cmd.get("alias")]

    # Finding all Vulkan types to be use by is_descendant_type
    types = {}
    for type in spec.findall("types/type"):
        name = type.findtext("name")
        if name:
            types[name] = type

    for key in block_keys:
        blocks[key] = ""

    # For each group, get the list of all commands
    for group, cmdnames in command_groups.items():
        # Skipping some extensions
        if group in ExcludeList:
            continue

        ifdef = "#if " + group + "\n"

        for key in block_keys:
            blocks[key] += ifdef

        # Name the NVVK_HAS with the first part of the group
        ext_name = group
        if "&&" in group:
            ext_name = group.split("&&")[0].strip()
        elif "||" in group:
            ext_name = group.split("||")[0].strip()
        if ext_name != None:
            blocks["DEFINE"] += "#define NVVK_HAS_" + remove_defined(ext_name) + "\n"

        # Getting all commands within the group
        for name in sorted(cmdnames):
            # Finding the 'alias' command
            cmd = commands[name]

            if name in ExportedCommands:
                print("Command " + name + " from group " + group)

            # Get the first argument type, which defines if it is an instance
            # function
            type = cmd.findtext("param[1]/type")

            # Create the function declaration block
            params = cmd.findall("param")
            return_type = cmd.findtext("proto/type")
            blocks["DECLARE"] += get_function(return_type, name, params)

            # Loading proc address can be device or instance
            if (
                is_descendant_type(types, type, "VkDevice")
                and name not in instance_commands
            ):
                blocks["LOAD_PROC"] += (
                    "  pfn_"
                    + name
                    + " = (PFN_"
                    + name
                    + ')getDeviceProcAddr(device, "'
                    + name
                    + '");\n'
                )
            elif is_descendant_type(types, type, "VkInstance"):
                blocks["LOAD_PROC"] += (
                    "  pfn_"
                    + name
                    + " = (PFN_"
                    + name
                    + ')getInstanceProcAddr(instance, "'
                    + name
                    + '");\n'
                )

            # Creates the bloc for all static functions
            blocks["STATIC_PFN"] += "static PFN_" + name + " pfn_" + name + "= 0;\n"

        # Adding the #endif or removing empty blocks
        for key in block_keys:
            if blocks[key].endswith(ifdef):
                blocks[key] = blocks[key][: -len(ifdef)]
            else:
                blocks[key] += "#endif /* " + remove_defined(group) + " */\n"

    # Patching the files
    patch_file("extensions_vk.hpp", blocks)
    patch_file("extensions_vk.cpp", blocks)
